### PR TITLE
Refactor Wait Times

### PR DESCRIPTION
**Pull request by whoiswillma**
#163 

output

### CODE REVIEW

File: Eatery/Models/CampusEatery.swift [hunk 1]
```diff
@@ -113,7 +124,9 @@ struct CampusEatery: Eatery {
         }
 
         self.allEvents = events.flatMap { $0.value.map { $0.value } }
-        self.swipeData = swipeData
+        self.swipeDataByHour = swipeDataPoints.reduce(into: [:], { (swipeDataByHour, point) in
+            swipeDataByHour[point.militaryHour, default: []].insert(point)
+        })
```
Comment (kevinchan159) [inline Eatery/Models/CampusEatery.swift:129]: clean!

File: Eatery/Views/Eateries/PopularTimesView.swift [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+    private var histogramView = HistogramView()
+
+    private let startHour = 6
+    private let overflowedEndHour = 27
+
+    init(eatery: CampusEatery) {
+        self.eatery = eatery
+
+        super.init(frame: .zero)
```
Comment (kevinchan159) [inline Eatery/Views/Eateries/PopularTimesView.swift:51]: can we put `super.init(.zero)` as the first line in the function?

File: Eatery/Views/Eateries/PopularTimesView.swift [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        setExpanded(!isExpanded, animated: true)
+    }
+
+    func setExpanded(_ newValue: Bool, animated: Bool) {
+        isExpanded = newValue
+        if isExpanded {
+            showHideButton.setTitle("Hide", for: .normal)
+        } else {
+            showHideButton.setTitle("Show", for: .normal)
+        }
```
Comment (kevinchan159) [inline Eatery/Views/Eateries/PopularTimesView.swift]: can we use ternary here?
```
let title = isExpanded ? "Hide" : "Show"
showHideButton.setTitle(title, for: .normal)
```

File: Eatery/Views/Eateries/PopularTimesView.swift [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            [NSAttributedStringKey.foregroundColor: UIColor.eateryBlue,
+             NSAttributedStringKey.font: UIFont.boldSystemFont(ofSize: 14)
+            ],
+            range: NSRange(location: hourText.count, length: waitTimesText.count)
+        )
+        return string
+    }
+
+    func numberOfDataPointsPerTickMark(for histogramView: HistogramView) -> Int? {
+        return 3
```
Comment (kevinchan159) [inline Eatery/Views/Eateries/PopularTimesView.swift:193]: can we add comment about why this is 3?

**Review by kevinchan159 — APPROVED** [review]
WOW THIS IS SO COOL!!! 🤩

File: Eatery/Controllers/Eateries/CampusEateryMenuViewController.swift [hunk 5]
```diff
@@ -513,3 +456,11 @@ extension CampusEateryMenuViewController: TabbedPageViewControllerDelegate {
...
 }
 
+extension CampusEateryMenuViewController: PopularTimesViewLayoutDelegate {
+
+    func popularTimesContentSizeDidChange(_ popularTimesView: PopularTimesView) {
+        outerScrollView.layoutIfNeeded()
+    }
+
+}
+
```
Comment (young-k) [inline Eatery/Controllers/Eateries/CampusEateryMenuViewController.swift]: extra newline?

File: Eatery/Views/Eateries/PopularTimesView.swift [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+ Why is this layout delegate needed?
+
+ To animate the expansion of the popular times view, we must change the frame
+ of the scroll view. The popular times view is able to manage its own
+ constraints to change its own height. However, without direct access to the
+ outer scroll view (in CampusEateryMenuViewController), it is unable to
+ cleanly animate the change.
+
+ This delegate is the cleanest (albeit still not super clean) way to access the
+ outer scroll view.
```
Comment (young-k) [inline Eatery/Views/Eateries/PopularTimesView.swift:22]: really minor, but isn't the normal convention:

```
/**
 * asdf
 *
 */
```

File: Eatery/Views/Eateries/PopularTimesView.swift [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        let currentHour = Calendar.current.component(.hour, from: Date())
+        let barHour = containOverflowedMilitaryHour(hour: startHour + index)
+
+        let hourText = currentHour == barHour ? "Now: " : formattedHourForTime(militaryHour: barHour).appending(": ")
+
+        let waitTimesText: String
+        if let (low: lowEstimate, high: highEstimate) = eatery.waitTimes(atHour: barHour, minute: 0) {
+            waitTimesText = "\(lowEstimate)-\(highEstimate)m"
+        } else {
+            waitTimesText = "?"
```
Comment (young-k) [inline Eatery/Views/Eateries/PopularTimesView.swift:174]: not sure if this should be `?`, does design have a state for this?

File: Eatery/Views/Eateries/PopularTimesView.swift (hunk 2 above)
Comment (whoiswillma) [reply Eatery/Views/Eateries/PopularTimesView.swift:51]: The `eatery` property must be initialized before `super.init` is called, so no.

File: Eatery/Views/Eateries/PopularTimesView.swift (hunk 7 above)
Comment (whoiswillma) [reply Eatery/Views/Eateries/PopularTimesView.swift:174]: I'll ask

File: Eatery/Views/Eateries/PopularTimesView.swift (hunk 6 above)
Comment (whoiswillma) [reply Eatery/Views/Eateries/PopularTimesView.swift:22]: Hmmn, I did a bit of looking and apparently Apple uses `///` on every line to do their documentation. 

e.g. https://github.com/apple/swift/blob/master/stdlib/public/core/String.swift

Xcode parses the documentation cleanly regardless of the style of comment. Any thoughts on standardized comments? I don't think it matters all that much; it's consistency for consistency's sake.

File: Eatery/Views/Eateries/PopularTimesView.swift (hunk 4 above)
Comment (whoiswillma) [reply Eatery/Views/Eateries/PopularTimesView.swift:193]: I can't think of any comment that isn't a restatement of the method name?

**Review by ReadeP — APPROVED** [review]
🔥🔥🔥

**Pull request merged**